### DISCUSSION TITLE
How to constrain an affine transform?

### DISCUSSION THREAD

**Discussion by raacampbell** [SuperElastix/elastix#875] [Q&A]
I am registering two 3D images. Due to the way the images are generated I know only certain transformations are physically possible. e.g. I can have translation in Y and Z but not X. Differences in scale and rotation between my images are not possible. 

My results are not as good as expected and the main thing going wrong at the moment is that Elastix is calculating a non-zero X translation. I verified this by manually setting this parameter to zero and re-transforming the image with `transformix`. 

Is there any way I can constrain my registration? Presumably I will get a better result this way compared to setting parameters to zero *post hoc*.

**Comment by ntatsisk** [comment]
Hi @raacampbell, what I would try is to set the `Scales` of the directions that you want to constrain to zero. My understanding is that `Scales`  are the weights/factors that correspond to each parameter during optimization. You can search the word here to get a better idea: https://elastix.lumc.nl/doxygen/parameter.html. I also quote from the manual at 5.3.5 section:

> Another parameter that needs to be set is the Scales.
> The scales define for each element of the transformation parameters /mu a scaling value, which is used during
> optimisation. The scaling serves to bring the elements of /mu in the same range (parameters corresponding to
> rotation have in general a much smaller range than parameters corresponding to translation). We recommend
> to let elastix compute it automatically: (AutomaticScalesEstimation "true")1. Always start with a
> rigid or affine transformation before doing a nonrigid one, to get a good initial alignment.

I would try a "dry-run" with maximum iterations =1 and (AutomaticScalesEstimation "true"), then I would copy the `Scales` from the result TransformParameters, set to zero the directions to be constrained and re-run the registration with  (AutomaticScalesEstimation "false").

That is just an idea though. Maybe somebody else has a better one or has tried it before.

**Comment by raacampbell** [comment]
I tried scales and it seemed to make no difference but I might have done it wrong. What did help, however, was translation only reg (not affine). I still need to set to 0 one of the parameters but overall the reg works better even without that.

**Comment by ntatsisk** [comment]
I created a notebook example on how to use scales for that purpose (https://github.com/InsightSoftwareConsortium/ITKElastix/blob/main/examples/ITK_Example21_ConstrainedRegistration.ipynb). @raacampbell, could you take a look?
